internal/assetapi/assetapi.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2025 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package assetapi
+
+import (
+	"fmt"
+	"os/exec"
+
+	"github.com/palantir/pkg/safejson"
+	"github.com/pkg/errors"
+)
+
+type AssetType string
+
+const (
+	ConjureIRExtensionsProvider AssetType = "conjure-ir-extensions-provider"
+)
+
+type LoadedAssets struct {
+	ConjureIRExtensionsProviders []string
+}
+
+// LoadAssets takes a list of asset paths and returns a LoadedAssets struct that contains the typed assets. Returns an
+// error if any of the provided assets are not valid according to the plugin asset specification.
+func LoadAssets(assets []string) (LoadedAssets, error) {
+	loadedAssets, err := loadAssets(assets)
+	if err != nil {
+		return LoadedAssets{}, err
+	}
+	return LoadedAssets{
+		ConjureIRExtensionsProviders: loadedAssets[ConjureIRExtensionsProvider],
+	}, nil
+}
+
+// loadAssets takes a list of asset paths, determines their types, and returns a map from AssetType to the list of
+// assets of that type. Returns an error if any asset cannot be executed to determine its type, or if any asset is of an
+// unsupported type.
+func loadAssets(assets []string) (map[AssetType][]string, error) {
+	assetMap := make(map[AssetType][]string)
+	for _, asset := range assets {
+		assetType, err := getAssetTypeForAsset(asset)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get asset type for asset %s", asset)
+		}
+
+		if assetType != ConjureIRExtensionsProvider {
+			return nil, fmt.Errorf("unsupported asset type %s for asset %s: only asset type that is supported is %q", assetType, asset, ConjureIRExtensionsProvider)
+		}
```

Comment:
This is new behavior: enforcing that, even if `type` field is present, loading will fail unless type is recognized.

However, adding a new asset type is fairly straightforward.

This pattern is largely analogous to existing asset loading in `distgo` etc.